httputil/cert_ext_tripper.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package httputil
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/asn1"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+type ignoreExtensionsRoundTripper struct {
+	base         *http.Transport
+	extsToIgnore []asn1.ObjectIdentifier
+	logger       hclog.Logger
+}
+
+// NewIgnoreUnhandledExtensionsRoundTripper creates a RoundTripper that may be used in an HTTP client which will
+// ignore the provided extensions if present unhandled on a certificate.  If base is nil, the default RoundTripper is used.
+func NewIgnoreUnhandledExtensionsRoundTripper(logger hclog.Logger, base http.RoundTripper, extsToIgnore []asn1.ObjectIdentifier) http.RoundTripper {
+	if len(extsToIgnore) == 0 {
+		return base
+	}
+	if base == nil {
+		base = http.DefaultTransport
+	}
+
+	tp, ok := base.(*http.Transport)
+	if !ok {
+		// We don't know how to deal with this object, bail
+		return base
+	}
+
+	return &ignoreExtensionsRoundTripper{base: tp, logger: logger, extsToIgnore: extsToIgnore}
+}
+
+func (i *ignoreExtensionsRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	var domain string
+	if strings.ContainsRune(request.URL.Host, ':') {
+		var err error
+		domain, _, err = net.SplitHostPort(request.URL.Host)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse domain from URL host %s", request.URL.Host)
+		}
+	} else {
+		domain = request.URL.Host
+	}
+
+	// Only update our values if the end-user hasn't overridden anything we wanted to do.
+	var perReqTransport *http.Transport
+	if !i.base.TLSClientConfig.InsecureSkipVerify && i.base.TLSClientConfig.VerifyConnection == nil {
```

Comment:
Checking for these could be done in the constructor as they aren't going to change from call to call, and with an appropriate error.